lua/barbar/layout.lua

```diff
@@ -20,130 +20,159 @@ local state = require'barbar.state'
 --- The number of sides of each buffer in the tabline.
 local SIDES_OF_BUFFER = 2
 
+--- The length of a slash (`#'/'`)
+local SLASH_LEN = #'/'
+
+--- The length of one space (`#' '`)
+local SPACE_LEN = #' '
+
 --- @class barbar.layout.data
---- @field actual_width integer
---- @field available_width integer
---- @field base_widths integer[]
---- @field buffers_width integer
---- @field padding_width integer
---- @field tabpages_width integer
---- @field used_width integer
+--- @field actual_width integer the sum of the `base_widths` plus the `padding_width` allocated to each buffer
+--- @field base_widths integer[] the minimum amount of space taken up by each buffer
+--- @field buffers_width integer the amount of space available to be taken up by buffers
+--- @field padding_width integer the amount of padding used on each side of each buffer
+--- @field scroll_max integer the maximum position which can be scrolled to
+--- @field tabpages_width integer the amount of space taken up by the tabpage indicator
 
 --- @class barbar.Layout
 --- @field buffers integer[] different from `state.buffers` in that the `hide` option is respected. Only updated when calling `calculate_buffers_width`.
 local Layout = {buffers = {}}
 
---- The number of characters needed to represent the tabpages.
---- @return integer width
-function Layout.calculate_tabpages_width()
-  local current = tabpagenr()
-  local total   = tabpagenr('$')
-  if not config.options.tabpages or total == 1 then
-    return 0
-  end
-  return 1 + tostring(current):len() + 1 + tostring(total):len() + 1
+--- Calculate the current layout of the bufferline.
+--- @return barbar.layout.data
+function Layout.calculate()
+  local available_width = get_option'columns'
+  available_width = available_width - state.offset.left.width - state.offset.right.width
+
+  local pinned_count, pinned_sum, sum, widths = Layout.calculate_buffers_width()
+  local tabpages_width = Layout.calculate_tabpages_width()
+
+  local pinned_width = pinned_sum + (pinned_count * config.options.minimum_padding * SIDES_OF_BUFFER)
+  local buffers_width = available_width - pinned_width - tabpages_width
+  local count = #widths - pinned_count
+
+  local remaining_width = max(0, buffers_width - sum)
+  local remaining_width_per_buffer = floor(remaining_width / count)
+  local remaining_padding_per_buffer = floor(remaining_width_per_buffer / SIDES_OF_BUFFER)
+  local padding_width = max(config.options.minimum_padding, min(remaining_padding_per_buffer, config.options.maximum_padding))
+  local actual_width = sum + (count * padding_width * SIDES_OF_BUFFER)
+
+  return {
+    actual_width = actual_width,
+    base_widths = widths,
+    buffers_width = buffers_width,
+    padding_width = padding_width,
+    scroll_max = max(0, actual_width - buffers_width),
+    tabpages_width = tabpages_width,
+  }
```

Comment:
If you find better names, it's also ok to refactor. I was re-reading this the other day and I feel like some of this isn't 100% clear.

Reply:
Sounds good. I ended up documenting all the meanings of these, so I'll see if I can come up with a clearer name.